Bump Eigen to 3.4.0

DavidFair opened the pull request:
**Description of work.**
Bumps Eigen to the 3.4 series. We could stick with 3.3.9 to fix
deprecated C++-17 warnings, but as there is only one breaking change
with 3.4.0 lets go all the way up in one single hit

**To test:**
- Check that all tests pass

There is no associated issue. - Fixes an issue seen on CI with newer MSVC compiler

*This does not require release notes* because it's not a user visible change

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Bump Eigen to 3.4.0

Bumps Eigen to the 3.4 series. We could stick with 3.3.9 to fix
deprecated C++-17 warnings, but as there is only one breaking change
with 3.4.0 lets go all the way up in one single hit

KedoKudo reviewed (dismissed):
simple version bump, straight forward enough

DavidFair committed:
Remove option to use system Eigen

Remove the option to use System Eigen so we get nice reproducable
builds, especially as we switch to Conda.

This means we implicitly opt MSVC clients into building Eigen rather
than using a pre-built version

DavidFair commented:
@KedoKudo / @martyngigg Could you look at this again

The original commit wouldn't fix MSVC as it uses a pre-built. I've shifted that to now use the external project version rather than our third party deps repo

DavidFair committed:
Modernise CMake for Eigen

Switches from externalproject_add to fetch content.
On MSVC this fails to identify a C++03/11 compiler during configure, but
because it's a header-only lib we can still compile with it regardless

DavidFair commented:
It looks like it's picking up the Qt4 backend on our OSX builders, any objections to me nuking the dependency from brew?

martyngigg commented:
I think the Qt4 errors are warnings? I agree we don't need it but I'm a little loathed to fiddle with those machines too much.

The error seems to something about the root variable:

```
18:35:39 CMake Error at /usr/local/Cellar/cmake/3.17.3/share/cmake/Modules/CMakePackageConfigHelpers.cmake:265 (message):
18:35:39   Variable EIGEN_ROOT_DIR does not exist
```

Should we just upgrade the internal version inside the thirdparty instead for now?

DavidFair committed:
Fix CMake defaulting to /usr/local for the install directory

OSX was failing as it didn't know where to install the headers to. More fundamentally we
shouldn't be polluting the system with Mantid headers, so manually set the location
before calling FetchContent_MakeAvailable

DavidFair commented:
Turns out it was a PEBKAC problem:
We were attempting to install Eigen to system directories on all platforms. OSX just does a better job protecting system dirs, hence the job failed

DavidFair committed:
Suppress unknown warning on older XCode clang

Our CI servers don't understand -Wimplicit-int-float-conversion, so eigen trying to
suppress it emits further warnings. We can simply disable warnings about unknown warnings
(rabbit holes all the way down) to make everything work

DavidFair committed:
Suppress -WConversion in upstream Eigen

Hyperplane uses an alignment macro in xattr which triggers a conversion
from unsigned int to int. Suppress this until it's fixed upstream (as of
this commit master still shows the same warning)

DavidFair committed:
Stop cppcheck checking Eigen Dependencies

Since we have moved the directory to _deps we now are picking them up
again. Simply mark everything in the build dir off-limits to cppcheck

DavidFair committed:
Remove unused suppressions from cppcheck

Adding MantidKernel/SuppressWarnings apparently will cause cppcheck to
skip the downstreams cpp and .h files. This means that everything that
implicitly now has the warning suppression is no longer flagging
warnings at all.

As this has been a long-standing problem within the project, I suspect
we have a number of files with missed warnings because of this. For that
reason I think this commit should stand, and a future series should fix
the suppression then re-instate these warnings plus all the other ones
in the project

DavidFair commented:
This should be ready to go now :crossed_fingers: 
My hope is by making it the same everywhere it makes it easier when we switch to Conda later on / upgrading it in the future too

DavidFair commented on buildconfig/CMake/Eigen.cmake:
```diff
@@ -1,30 +1,37 @@
-  set(USE_SYSTEM_EIGEN ON)
-endif()
+# Preserve shared attrs
+set(CMAKE_INSTALL_PREFIX_OLD ${CMAKE_INSTALL_PREFIX})
```
Both libs install to `/usr/local` on Linux. We don't call the install target on any of these 3 deps, but Eigen uses the install_directory to configure some variables (hence `EIGEN_ROOT_DIR` was unset) just in-case we do run the install target

Agree 100%, I'd like to drop span (as I'm hoping the Conda compilers will have C++-20 support) and move Eigen and GTest over
My hope is by bringing all platforms over to 3.4.0 it makes it trivial for us to simply migrate this dep over as all platforms are already running the same version. If long-term we were going to stick with the third-party repo and *Nix system libs I'd have done this in the MSVC deps repo instead and dropped this PR

Pasarus replied:
The conda compilers are for only Linux, I see few issues with moving to GCC 11 for example on Linux, we will be dependent on MSVC and XCode versions moving forward with C++ standard support.

martyngigg reviewed (approved):
I'm happy with this now. It builds for me locally.

Just one question but not something to hold this up. Is it possible to tell Eigen we don't want to look for Qt?

DavidFair commented:
So grepping their repo: `EIGEN_TEST_NOQT`, though it only affects a doc and test for a Mandelbrot set that requires Qt4
Want me to throw another commit on to pass this in too?

martyngigg reviewed (changes requested):
I think that would be good.

Are the options actually taking effect? I notice that the testing option is off so I would have expected that `find_package(Qt4)` not even to run?

DavidFair committed:
Disable Qt in Eigen builds

We dont use the Qt backend for anything, and it's only adding potential
problems long-term

DavidFair commented:
Done, and it still checks for configuring the `demo` and `doc` targets
It's worth double checking our OSX build if it tries to skip Qt4, as I don't have this installed locally

martyngigg commented:
Yeah it's not worth holding these fixes up any further. I don't quite understand what's going on but using an available Eigen build is a better long term solution anyway.

Pasarus reviewed (changes requested):
Could you bump this in the places for conda as well please? mantid-developer yml files would be first, and second would be the conda-recipes, in the conda-recipes repo.

martyngigg reviewed (changes requested):
@thomashampson  raised the issue of the cppcheck suppression removal and I think we do need to understand that or as ever it will just get left.

The [manual](https://cppcheck.sourceforge.io/manual.pdf) says that once macros `#if` are introduced then cppcheck will behave differently. Are we now seeing behaviour where cppcheck is just skipping run code because it has either maxed out its config depth or it's not running through a branch of code?

I wondered what might happen if we say forced cppcheck to look like `gcc` and using that config?

DavidFair committed:
Bump Eigen in Conda dependencies

Bumps Eigen to 3.4.x for Conda builds

martyngigg commented:
I've had a brief look/think about this and a couple of things.

Firstly I spotted that we should probably add `-i "${CMAKE_BINARY_DIR}"` to the `cppcheck` call as this will ignore those files and they won't even be checked. That way we don't need to suppressions and it speeds up the check as it skips that directory entirely.

One downside of switching to `FetchContent` is that `Eigen3` is not not treated as a system include so we have to manually suppress the warnings. I can't find a way to fix this as it suggests that `FetchContent` is really for things you want to consider part of your project. Is `ExternalProject` not a better model here as it is external and when we get an imported target it comes with `-isystem` so we don't need to manually suppress the warnings?

The only other thing would then be the `USE_SYSTEM_EIGEN` flag - do we not need this when we build with conda as we really want to use that anyway?

DavidFair commented:
> Firstly I spotted that we should probably add `-i "${CMAKE_BINARY_DIR}"` to the `cppcheck` call as this will ignore those files and they won't even be checked. That way we don't need to suppressions and it speeds up the check as it skips that directory entirely.

Good point, I'll open a separate PR for this in case this one stalls out [Link TODO]

> Is `ExternalProject` not a better model here as it is external and when we get an imported target it comes with `-isystem` so we don't need to manually suppress the warnings?

I ended up switching these because `ExternalProject` obtains the source at build time (I'd normally link to the CMake docs paragraph, but there's no useful in-line link on that page), whereas we need them at configure time. The old mechanism hacked around this by simply invoking CMake to perform a "build" of Eigen during the configuration. 
This adds time onto Windows as we now have spin up CMake 3 times (Mantid, Eigen Configure and Eigen build/download). Whereas setting it to FetchContent makes it available at configure time in the top level CMake directly.

I wasn't aware of that, we can do this by [overriding the properties](https://stackoverflow.com/a/64139281) after calling FetchContent, I'll try that fix and see if we get `WConversion`....

> The only other thing would then be the `USE_SYSTEM_EIGEN` flag - do we not need this when we build with conda as we really want to use that anyway?
Woops I read `SYSTEM_EIGEN` as not one we control, forgetting that Conda appears like a system dependency

DavidFair committed:
Fix Eigen Conda support + mark as system include

Fixes Eigen to skip trying to clone the dep on Conda (long-term we can
simply remove this entire file)
Marks the entire package as a system include directory to suppress
warnings

DavidFair committed:
Revert "Suppress -WConversion in upstream Eigen"

This reverts commit af4e5732d5c176bc41def77e3df7db9804ba6cdb.
Marking the include directories as system should suppress warnings
automatically

DavidFair committed:
Revert "Stop cppcheck checking Eigen Dependencies"

This reverts commit a615164209531a98f5c08eb50e3845486dea4ee6.

DavidFair committed:
Revert "Remove unused suppressions from cppcheck"

This reverts commit babb85255b09978d82248b43ff417e47255f01bb.

martyngigg commented on buildconfig/CMake/Eigen.cmake:
```diff
@@ -1,30 +1,44 @@
-  execute_process(COMMAND ${CMAKE_COMMAND} --build . WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/extern-eigen)
+  # Mark target as a system include
+  get_target_property(EIGEN_INC_DIRS eigen INTERFACE_INCLUDE_DIRECTORIES)
+  set_target_properties(eigen PROPERTIES INTERFACE_SYSTEM_INCLUDE_DIRECTORIES "${EIGEN_INC_DIRS}")
```
Cool. Nice that we can keep the benefits of not running cmake again and again but keeping the system include 👍🏻

martyngigg reviewed (approved):
I'm happy with the changes here as they now only affect how we grab Eigen but have improved the performance of that mechanism on Windows.

martyngigg commented:
@Pasarus Could you check the conda stuff here and then we could get this merged.

martyngigg commented:
@mantidproject/gatekeepers Could I get this merged?
